indra/databases/cbio_client.py

```diff
@@ -88,12 +89,92 @@ def get_mutations(study_id, gene_list, mutation_type=None,
+                             'mutation_type', mutation_type)
     mutations = {'gene_symbol': list(res['gene_symbol'].values()),
                  'amino_acid_change': list(res['amino_acid_change'].values())}
     return mutations
 
 
+def get_case_lists(study_id):
+    """Return a list of the case set ids for a particular study.
+
+    TAKE NOTE the "case_list_id" are the same thing as "case_set_id"
+    Within the data, this string is referred to as a "case_list_id".
+    Within API calls it is referred to as a 'case_set_id'.
+    The documentation does not make this explicitly clear.
+
+    Parameters
+    ----------
+    study_id : str
+        The ID of the cBio study.
+        Example: 'cellline_ccle_broad' or 'paad_icgc'
+
+    Returns
+    -------
+    case_set_ids : dict[dict[int]]
+        A dict keyed to cases containing a dict keyed to genes
+        containing int
+    """
+    data = {'cmd': 'getCaseLists',
+            'cancer_study_id': study_id}
+    df = send_request(**data)
+    case_set_ids = df['case_list_id'].tolist()
+    return case_set_ids
+
+
+def get_profile_data(study_id, gene_list, profile_filter, case_set_id=None):
+    """Return dict of cases and genes and their respective values.
+
+    Parameters
+    ----------
+    study_id : str
+        The ID of the cBio study.
+        Example: 'cellline_ccle_broad' or 'paad_icgc'
+    gene_list : list[str]
+        A list of genes with their HGNC symbols.
+        Example: ['BRAF', 'KRAS']
+    profile_filter : str
+        A string used to filter the profiles to return. Will be one of:
+        - MUTATION
+        - MUTATION_EXTENDED
+        - COPY_NUMBER_ALTERATION
+        - MRNA_EXPRESSION
+        - METHYLATION
+    case_set_id : Optional[str]
+        A string that specifices which case_set_id to use, based on a complete
+        or partial match. If not provided, will look for study_id + '_all'
+
+    Returns
+    -------
+    profile_data : dict[dict[int]]
+        A dict keyed to cases containing a dict keyed to genes
+        containing int
+    """
+    genetic_profile = get_genetic_profiles(study_id, profile_filter)[0]
+    gene_list_str = ','.join(gene_list)
+    case_set_ids = get_case_lists(study_id)
+    if case_set_id:
+        case_set_id = [x for x in case_set_ids if case_set_id in x][0]
+    else:
+        case_set_id = study_id + '_all'
+        # based on looking at the cBioPortal, this is a common case_set_id
+    data = {'cmd': 'getProfileData',
+            'case_set_id': case_set_id,
+            'genetic_profile_id': genetic_profile,
+            'gene_list': gene_list_str,
+            'skiprows': 2}
+    df = send_request(**data)
+    case_list_df = [x for x in df.columns.tolist()
+                    if x not in ['GENE_ID', 'COMMON']]
+    gene_list_df = df['COMMON'].tolist()
+    profile_data = _recursive_dict()
+    for case in case_list_df:
+        profile_values = df[case].tolist()
```

Comment:
This produces `nan` for missing values, I think it would be better to have `None` added to the result for missing values.

Reply:
Refactored code to use dict comprehension instead of `_recursive_dict()` function. In addition, it now produces `None` instead of `nan` for missing values. Commit b52febe